timewarp_gl/plugin.cpp

```diff
@@ -74,19 +74,19 @@ class timewarp_gl : public threadloop {
 	rendered_frame frame;
 
 	// Switchboard plug for application eye buffer.
-	std::unique_ptr<reader_latest<rendered_frame>> _m_eyebuffer;
+	switchboard::reader<rendered_frame> _m_eyebuffer;
 
 	// Switchboard plug for sending hologram calls
-	std::unique_ptr<writer<hologram_input>> _m_hologram;
+	switchboard::writer<switchboard::event_wrapper<std::size_t>> _m_hologram;
 
 	// Switchboard plug for publishing vsync estimates
-	std::unique_ptr<writer<time_type>> _m_vsync_estimate;
+	switchboard::writer<switchboard::event_wrapper<time_type>> _m_vsync_estimate;
 
 	// Switchboard plug for publishing MTP metrics
-	std::unique_ptr<writer<std::chrono::duration<double, std::nano>>> _m_mtp;
+	switchboard::writer<switchboard::event_wrapper<std::chrono::duration<double, std::nano>>> _m_mtp;
 
 	// Switchboard plug for publishing frame stale-ness metrics
-	std::unique_ptr<writer<std::chrono::duration<double, std::nano>>> _m_frame_age;
+	switchboard::writer<switchboard::event_wrapper<std::chrono::duration<double, std::nano>>> _m_frame_age;
```

Comment:
Why use `event_wrappers` for these specific writers? The writer in, for example, offline_imu_cam, is a "normal" writer without an event wrapper.
On the receiving end of these writes, the readers have to use a double dereference `**` instead of just a normal read. For example, `pose_lookup/plugin.cpp:61`.

Reply:
Events have to inherit from `switchboard::event`, so they have a known virtual destructor, so that memory can be cleaned up in switchboard. Any type (call it `T`) that we did not create in `data_format.hpp` then have to be "wrapped" in a class that inherits from `switchboard::event` and holds an inner `T`. `switchboard::event_wrapper<T>` is a short-hand for this.